### PR TITLE
granular test selection, take 3

### PR DESCRIPTION
**Pull request by garyverhaegen-da**
This is the final bit extracted from #6314. This PR adds the capability to filter individual test cases (for either inclusion or exclusion) while being completely backwards compatible. It also introduces a new cli option, `--list-all`, to list all test cases, keeping the behaviour of the existing `--list` option unchanged, i.e. only print test suites.

Unlike #6314, this is still building upon the idea of defining test suites as maps, preserving the known-good order of tests and thereby not introducing flakiness into our builds.

CHANGELOG_BEGIN

- [Ledger API Test Tool] ``--exclude`` and ``--include`` now match the
  full test name as a prefix, rather than just suite names. Test name is
built by combining the suite name with a test identifier, so this change
should be fully backwards compatible. Run with ``--list-all`` to list
all tests (as opposed to just the test suites with ``--list``).

CHANGELOG_END

### CODE REVIEW

File: docs/source/tools/ledger-api-test-tool/index.rst [hunk 1]
```diff
@@ -78,10 +78,36 @@ Tests that either change the global state of the ledger (e.g. configuration mana
...
-- ``--list``: print all available tests to the console, shows if they are run by default
-- ``--include``: only run the tests provided as argument
-- ``--exclude``: do not run the tests provided as argument
-- ``--all-tests``: run all default and optional tests. This flag can be combined with the ``--exclude`` flag.
+- ``--list``: print all available test suites to the console, shows if they are run by default
+- ``--list-all``: print all available tests to the console, shows if they are run by default
+- ``--include``: only run the tests that match the argument
+- ``--exclude``: do not run the tests that match the argument
+- ``--all-tests``: include all default and optional tests. This flag can be combined with the ``--exclude`` flag.
+- ``--perf-tests``: list performance tests to run; cannot be combined with normal tests
```
Comment (garyverhaegen-da) [inline docs/source/tools/ledger-api-test-tool/index.rst:86]: `--perf-tests` is not introduced by this PR, but was not documented before.

File: compatibility/bazel_tools/testing.bzl [hunk 2]
```diff
@@ -52,13 +52,26 @@ excluded_test_tool_tests = [
     },
     {
         "start": "1.1.1",
+        "end": "0.0.0",  #FIXME: change to 1.3 after release
```
Comment (garyverhaegen-da) [inline compatibility/bazel_tools/testing.bzl:55]: @cocreature is there a better way to handle this?

Comment (cocreature) [reply compatibility/bazel_tools/testing.bzl:55]: `0.0.0` is equivalent to just omitting `end`.

Comment (cocreature) [reply compatibility/bazel_tools/testing.bzl:55]: actually is this really what you want? The end is inclusive. I think you want the latest snapshot at the time of merging.

Comment (garyverhaegen-da) [reply compatibility/bazel_tools/testing.bzl:55]: > I think you want the latest snapshot at the time of merging.
Right. I'll update that before merging.

File: docs/source/tools/ledger-api-test-tool/index.rst [hunk 3]
```diff
@@ -78,10 +78,36 @@ Tests that either change the global state of the ledger (e.g. configuration mana
...
+
+Include and exclude are matched as prefixes, e.g. ``--exclude=SemanticTests``
+will exclude all tests whose name starts with ``SemanticTests``. Test names
+always start with their suite name followed by a colon, so ``--list`` can be
```
Comment (cocreature) [inline docs/source/tools/ledger-api-test-tool/index.rst:90]:
```suggestion
always start with their suite name followed by a colon, so the test suite names shown by ``--list`` can be
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala [hunk 4]
```diff
@@ -169,6 +169,11 @@ object Cli {
            |Party names must be their hints.""".stripMargin)
 
     opt[Unit]("list")
+      .action((_, c) => c.copy(listTestSuites = true))
+      .text(
+        """Lists all available test suites that can be used in the include and exclude options. Test names always start with their suite name, so using the suite name as a prefix matches all tests in a given suite.""")
+
+    opt[Unit]("list-all")
```
Comment (cocreature) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala:176]: Happy to leave this up to you but an alternative UX would be `list --all`. Not sure what I prefer.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala [hunk 5]
```diff
@@ -88,28 +104,25 @@ object LedgerApiTestTool {
...
+    val allTestCaseNames: Set[String] = cases(Tests.all(config)).map(_.name).toSet
+    val missingTests = (config.included ++ config.excluded).filterNot(prefix =>
+      allTestCaseNames.exists(_.startsWith(prefix)))
     if (missingTests.nonEmpty) {
-      println("The following tests could not be found:")
+      println("The following exclusion or inclusion does not match any test:")
       missingTests.foreach { testName =>
         println(s"  - $testName")
       }
       sys.exit(64)
```
Comment (cocreature) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala:119]: Totally unrelated but 64 seems like a weird choice :thinking:

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala [hunk 6]
```diff
@@ -130,13 +130,13 @@ object Cli {
       .action((ex, c) => c.copy(excluded = c.excluded ++ ex))
       .unbounded()
       .text(
-        """A comma-separated list of tests that should NOT be run. By default, no tests are excluded.""",
+        """A comma-separated list of exclusion prefixes. Tests whose name starts with any of the given prefixes will be skipped. Can be specified multiple times, e.g. `--exclude=a,b` is the same as `--exclude=a --exclude=b`.""",
```
Comment (oliverse-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala:133]:
```suggestion
        """A comma-separated list of exclusion prefixes. Tests whose name start with any of the given prefixes will be skipped. Can be specified multiple times, e.g. `--exclude=a,b` is the same as `--exclude=a --exclude=b`.""",
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala [hunk 7]
```diff
@@ -130,13 +130,13 @@ object Cli {
...
       )
 
     opt[Seq[String]]("include")
       .action((inc, c) => c.copy(included = c.included ++ inc))
       .unbounded()
-      .text("""A comma-separated list of tests that should be run.""")
+      .text("""A comma-separated list of inclusion prefixes. If not specified, all default tests are included. If specified, only tests that match at least one of the given inclusion prefixes (and none of the given exclusion prefixes) will be run. Can be specified multiple times, e.g. `--include=a,b` is the same as `--include=a --exclude=b`.""")
```
Comment (oliverse-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala:139]:
```suggestion
      .text("""A comma-separated list of inclusion prefixes. If not specified, all default tests are included. If specified, only tests that match at least one of the given inclusion prefixes (and none of the given exclusion prefixes) will be run. Can be specified multiple times, e.g. `--include=a,b` is the same as `--include=a --include=b`.""")
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala (hunk 5 above)
Comment (SamirTalwar-DA) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala:119]: I believe @rautenrieth-da picked it, based on [`sysexits`](https://man.openbsd.org/sysexits).

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala [hunk 8]
```diff
@@ -38,18 +38,29 @@ object LedgerApiTestTool {
   private def exitCode(summaries: Vector[LedgerTestSummary], expectFailure: Boolean): Int =
     if (summaries.exists(_.result.isLeft) == expectFailure) 0 else 1
 
-  private def printAvailableTests(config: Config): Unit = {
+  private def print[A](defaults: Seq[A], optionals: Seq[A])(t: A => String): Unit = {
```
Comment (stefanobaghino-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala:41]: The function signature is generic, but the implementation is quite specific. Mentioning the tests in the function name is more honest and memorable. Furthermore, it wouldn't shadow `print` from the `Prelude`.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala [hunk 9]
```diff
@@ -38,18 +38,29 @@ object LedgerApiTestTool {
...
+    optionals.map(t).sorted.foreach(println(_))
 
     println("\nAlternatively, you can run performance tests.")
     println(
       "Performance tests are not run by default, but can be run with `--perf-tests=TEST-NAME`.\n")
     Tests.PerformanceTestsKeys.sorted.foreach(println(_))
   }
+  private def printAvailableTestSuites(config: Config): Unit = {
+    println("Listing test suites. Run with --list-all to see individual tests.")
+    print(Tests.default.values.toSeq, Tests.optional(config).values.toSeq)(_.name)
```
Comment (stefanobaghino-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala:55]: Related to the shadowing I mentioned before, reading this in isolation would look a bit weird, given that `print` is defined in the `Prelude`.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerTestCase.scala [hunk 10]
```diff
@@ -9,12 +9,14 @@ import com.daml.lf.data.Ref
...
 
 final class LedgerTestCase(
+    val suite: LedgerTestSuite,
     val shortIdentifier: Ref.LedgerString,
     val description: String,
     val timeoutScale: Double,
     participants: ParticipantAllocation,
     runTestCase: ExecutionContext => Participants => Future[Unit],
 ) {
+  def name: String = s"${suite.name}:${shortIdentifier}"
```
Comment (stefanobaghino-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerTestCase.scala:19]:
```suggestion
  val name: String = s"${suite.name}:${shortIdentifier}"
```
Minor.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala (hunk 4 above)
Comment (garyverhaegen-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala:176]: I agree the `--list`/`--list-all` thing is a bit awkward, but I believe fixing this should be done as part of a broader redesign of the CLI; for example the situation with perf tests vs. normal tests could also use a more explicit structure. I don't want to extend the scope of this PR though.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala (hunk 8 above)
Comment (garyverhaegen-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala]: Renamed to `printListOfTests`.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala [hunk 11]
```diff
@@ -130,13 +130,13 @@ object Cli {
       .action((ex, c) => c.copy(excluded = c.excluded ++ ex))
       .unbounded()
       .text(
-        """A comma-separated list of tests that should NOT be run. By default, no tests are excluded.""",
+        """A comma-separated list of exclusion prefixes. Tests whose name start with any of the given prefixes will be skipped. Can be specified multiple times, e.g. `--exclude=a,b` is the same as `--exclude=a --exclude=b`.""",
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala:133]:
```suggestion
        """A comma-separated list of exclusion prefixes. Tests whose name start with any of the given prefixes will be skipped. Can be specified multiple times, i.e. `--exclude=a,b` is the same as `--exclude=a --exclude=b`.""",
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala [hunk 12]
```diff
@@ -130,13 +130,13 @@ object Cli {
...
       )
 
     opt[Seq[String]]("include")
       .action((inc, c) => c.copy(included = c.included ++ inc))
       .unbounded()
-      .text("""A comma-separated list of tests that should be run.""")
+      .text("""A comma-separated list of inclusion prefixes. If not specified, all default tests are included. If specified, only tests that match at least one of the given inclusion prefixes (and none of the given exclusion prefixes) will be run. Can be specified multiple times, e.g. `--include=a,b` is the same as `--include=a --include=b`.""")
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala:139]:
```suggestion
      .text("""A comma-separated list of inclusion prefixes. If not specified, all default tests are included. If specified, only tests that match at least one of the given inclusion prefixes (and none of the given exclusion prefixes) will be run. Can be specified multiple times, i.e. `--include=a,b` is the same as `--include=a --include=b`.""")
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala (hunk 4 above)
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala:176]: What happens if you use both `--list` and `--list-all`?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala [hunk 13]
```diff
@@ -38,18 +38,29 @@ object LedgerApiTestTool {
   private def exitCode(summaries: Vector[LedgerTestSummary], expectFailure: Boolean): Int =
     if (summaries.exists(_.result.isLeft) == expectFailure) 0 else 1
 
-  private def printAvailableTests(config: Config): Unit = {
+  private def printListOfTests[A](defaults: Seq[A], optionals: Seq[A])(t: A => String): Unit = {
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala:41]: Can we rename `t` to `getName` or something?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerTestSuite.scala [hunk 14]
```diff
@@ -24,7 +24,13 @@ private[testtool] abstract class LedgerTestSuite {
...
     val shortIdentifierRef = Ref.LedgerString.assertFromString(shortIdentifier)
     testCaseBuffer.append(
-      new LedgerTestCase(shortIdentifierRef, description, timeoutScale, participants, testCase),
+      new LedgerTestCase(
+        this,
+        shortIdentifierRef,
+        description,
+        timeoutScale,
+        participants,
+        testCase),
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerTestSuite.scala:33]:
```suggestion
        testCase,
      ),
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerTestCasesRunner.scala [hunk 15]
```diff
@@ -103,38 +103,37 @@ final class LedgerTestSuiteRunner(
...
   private def run(completionCallback: Try[Vector[LedgerTestSummary]] => Unit): Unit = {
-    val system: ActorSystem = ActorSystem(classOf[LedgerTestSuiteRunner].getSimpleName)
+    val system: ActorSystem = ActorSystem(classOf[LedgerTestCasesRunner].getSimpleName)
     implicit val materializer: Materializer = Materializer(system)
     implicit val executionContext: ExecutionContext = materializer.executionContext
 
     val participantSessionManager = new ParticipantSessionManager
     val ledgerSession = new LedgerSession(config, participantSessionManager)
-    val testCount = suites.map(_.tests.size).sum
+    val testCount = testCases.size
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerTestCasesRunner.scala:120]: Much nicer.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala (hunk 13 above)
Comment (garyverhaegen-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala:41]: Sure.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala (hunk 4 above)
Comment (garyverhaegen-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Cli.scala:176]: At the moment `--list` wins over everything. So if you say `--include=... --perf-tests=... --list` you still get the list of test suites. That's very much related to @cocreature's comment: we have a bunch of "flat" flags (i.e. no hierarchy) as far as the external interface is concerned, but internally we treat them as special things where some affect others. I think the tool has become rich enough that it's now worth having explicit subcommands, each which its own set of allowed flags.

**Comment by oliverse-da** [comment]
Excellent, @garyverhaegen-da , I am looking forward to start using the finer granularity.

So I take this will be part of today's 1.3-snapshot release?

**Comment by cocreature** [comment]
> So I take this will be part of today's 1.3-snapshot release?

Yep!

**Pull request merged**